clang/docs/checkedc/3C/INSTALL.md

```diff
@@ -31,10 +31,10 @@ you have already cloned one of these repositories, run the following
 # Get a copy of the Checked C system headers. Use Microsoft's
 # "checkedc" repository regardless of which "checkedc-clang"
 # repository you use.
-git clone https://github.com/microsoft/checkedc llvm/projects/checkedc-wrapper/checkedc
+git clone https://github.com/microsoft/checkedc checkedc-clang/llvm/projects/checkedc-wrapper/checkedc
```

Comment:
The document says above, "from the root of the working tree", by which I meant, "with the working directory set to `checkedc-clang`" (although maybe that wasn't clear).  These changes (1) seem to assume that the working directory is the parent of `checkedc-clang` instead of `checkedc-clang` itself and (2) will end up making the `build` directory a sibling of `checkedc-clang` instead of a child.  If you want to do (1), that needs to be made clear; alternatively, we could clarify the existing statement.  We can debate the merits of (2) if we want, though many of our tools already assume that `build` is a child and should be updated if we change our recommendation here.

Reply:
Sure. These changes were motivated by the set of commands you had in the 3c google doc which did not mention needing to run it in the working tree, and thus didn't build. I made the same change here so they were in sync. 
I'm not sure what build tool you're referring too, as I've had my tree setup like this for a long time without any issues.

Reply:
I guess the important decision is whether the `build` directory is a sibling of `checkedc-clang` or a child.  Once we decide that, we can figure out the clearest way to write the instructions both here and in the Google doc.  [Upstream LLVM's recommendation](https://www.llvm.org/docs/GettingStarted.html#getting-the-source-code-and-building-llvm) is a child, while [Microsoft's recommendation](https://github.com/microsoft/checkedc-clang/blob/master/clang/docs/checkedc/Setup-and-Build.md#setting-up-a-build-directory) (soon to be clarified in microsoft/checkedc-clang#1089) is a sibling: sigh.

Apparently, when I started at CCI in September 2020, I was aware of upstream LLVM's recommendation but not Microsoft's, so I went with the child layout.  I've since come to strongly prefer it: I use many linked working trees (`~/3c`, `~/3c-1.wt`, `~/3c-2.wt`, `~/3c-3.wt`, `~/3c-4.wt` on my laptop and several more on gamera) so I can leave several pending branches checked out at the same time and avoid having to switch branches and wait for rebuilds.  It would be hard to keep sibling build directories manageable without another level of directory nesting (`~/3c-N/src`, `~/3c-N/build`, etc.), which would make everything require more typing.  So in my personal scripts that I later contributed to `utils-private` (`llvm-run-tool` and `llvm-ninja`) and the test-updating programs (`clang/test/3C/testgenerator.py`, etc.), I chose to look for the build directory as a child of `checkedc-clang` if `LLVM_OBJ` is not set.  (My statement that "many of our tools already assume that `build` is a child" was too strong; it would be more accurate to say that "_several_ of our tools _favor_ a layout in which `build` is a child".)  Indeed, it's a reasonably safe assumption that if `checkedc-clang` contains a directory named `build`, that's the build directory, but it's less clear to me whether it would be reasonable to assume that if `checkedc-clang` has a sibling named `build`, that is the build directory; some additional configuration mechanism might be needed.

Of course, the needs of new users weigh heavily against those of the minority of power users like me, who will find the best way to get what they want from the available tools.

Are there countervailing convenience arguments (for users and/or scripts) in favor of `build` as a sibling?  Do we think consistency with our _nearest_ upstream (Microsoft) is the overriding consideration, or do we want to recommend the layout that we believe will be best for our users on the whole, with consistency with Microsoft potentially being one factor in that determination?  Or should we perhaps even try to convince Microsoft to recommend the child layout?

Regardless of which layout we recommend here, I assume that all our tools will continue to support the other layout in some way, via either specific automatic detection or some generic manual override like `LLVM_OBJ`.

(On a side topic re `LLVM_OBJ`: It looks like the only place the `LLVM_OBJ` environment variable was used before I arrived on the scene was in `convert_project`, so I assume I took it from there.  I'm not aware of any use of the `LLVM_OBJ` name outside CCI-written tools.  If there's another environment variable that's more standard, we could consider migrating to it.)

Reply:
Somehow the first time I built `clang`, I made `build` a sibling. I don't know why, I was following someone else's instructions. I'm persuaded by Matt's argument that build makes sense as a child directory, so I've been doing that for a while. If that's the LLVM/Clang recommendation, then we should do that. If Microsoft is not following that recommendation, I have a suspicion that it's due to the same inertia I experienced, and not for a particular/good reason, but maybe I'm wrong.

I observe that by default (I think) CLion puts the build directory within as a child directory (in clang/llvm/build).